qiskit/primitives/containers/data_bin.py

```diff
@@ -15,110 +15,138 @@
+    """Namespace for storing data.
 
-    def __repr__(cls):
-        name = cls.__name__
-        if cls._SHAPE is None:
-            return name
-        shape = ",".join(map(str, cls._SHAPE))
-        return f"{name}<{shape}>"
+    .. code-block:: python
 
+        data = DataBin(
+            alpha=BitArray.from_bitstrings(["0010"]),
+            beta=np.array([1.2])
+        )
 
-class DataBin(metaclass=DataBinMeta):
-    """Base class for data bin containers.
+        print("alpha data:", data.alpha)
+        print("beta data:", data.beta)
 
-    Subclasses are typically made via :class:`~make_data_bin`, which is a specialization of
-    :class:`make_dataclass`.
     """
 
-    _RESTRICTED_NAMES = {
-        "_RESTRICTED_NAMES",
-        "_SHAPE",
-        "_FIELDS",
-        "_FIELD_TYPES",
-        "keys",
-        "values",
-        "items",
-    }
-    _SHAPE: tuple[int, ...] | None = None
-    _FIELDS: tuple[str, ...] = ()
-    """The fields allowed in this data bin."""
-    _FIELD_TYPES: tuple[type, ...] = ()
-    """The types of each field."""
+    _RESTRICTED_NAMES = frozenset(
+        {
+            "_RESTRICTED_NAMES",
+            "_SHAPE",
+            "_FIELDS",
+            "_FIELD_TYPES",
+            "_data",
+            "_shape",
+            "keys",
+            "values",
+            "items",
+            "shape",
+            "ndim",
+            "size",
+        }
+    )
+
+    def __init__(self, *, shape: ShapeInput = (), **data):
+        """
+        Args:
+            data: Name/value data to place in the data bin.
+            shape: The leading shape common to all entries in the data bin. This defaults to
+                the trivial leading shape of ``()`` that is compatible with all objects.
+
+        Raises:
+            ValueError: If a name overlaps with a method name on this class.
+            ValueError: If some value is inconsistent with the provided shape.
+        """
+        if not self._RESTRICTED_NAMES.isdisjoint(data):
+            bad_names = sorted(self._RESTRICTED_NAMES.intersection(data))
+            raise ValueError(f"Cannot assign with these field names: {bad_names}")
+
+        _setattr = super().__setattr__
+        _setattr("_shape", shape_tuple(shape))
+        _setattr("_data", data)
+
+        ndim = len(self._shape)
+        for name, value in data.items():
+            if getattr(value, "shape", shape)[:ndim] != shape:
+                raise ValueError(f"The value of {name} does not lead with the shape {shape}.")
+            _setattr(name, value)
+
+        super().__init__()
 
     def __len__(self):
-        return len(self._FIELDS)
+        return len(self._data)
+
+    def __setattr__(self, *_):
+        raise NotImplementedError
 
     def __repr__(self):
-        vals = (f"{name}={getattr(self, name)}" for name in self._FIELDS if hasattr(self, name))
-        return f"{type(self)}({', '.join(vals)})"
+        vals = (f"{name}={val}" for name, val in self.items())
```

Comment:
This might be off-topic but I have a concern about repr of ndarray. Both `str(ndarray)` and `repr(ndarray)` print all matrix elements and it may break `repr(data_bin)` if the array size is large. On the other hand, BitArray displays a concise text such as `BitArray(<shape=(), num_shots=2, num_bits=4>)`.

It might be nice to introduce a specialized text format only for ndarray, e.g.,
- `f"<ndarray(<shape={array.shape}, dtype={array.dtype}>)>"`
  - `ndarray(<shape=(2, 3), dtype=bool>)`.

Reply:
Done in [d049863](https://github.com/Qiskit/qiskit/pull/12219/commits/d049863b601ec50a1e440c195d4fe9971d0ac9ab)

Reply:
Thanks. It looks great.
```
db = DataBin(a=np.empty((1, 2, 3)), shape=(2, 3))
print(db)
# DataBin(a=np.ndarray(<shape=(1, 2, 3), dtype=float64>), shape=(1, 2))
```